src/components/application_manager/src/application_data_impl.cc

```diff
@@ -582,6 +582,122 @@ void DynamicApplicationDataImpl::set_display_layout(const std::string& layout) {
   window_templates_[DEFAULT_WINDOW] = template_config;
 }
 
+/**
+ * @brief RemoveUnsupportedLayouts removes unsupported layouts from
+ * configurable_keys array, if present
+ * @param keyboard_capabilities Modified keyboard capabilities
+ * @param supported_layouts Set with all supported layouts, generated from
+ * supported_keyboard_layouts array
+ * @param customizable_layouts Set with all supported layouts, which keys may be
+ * customized (if not, number of keys for such layout will be set to zero by
+ * default)
+ */
+void RemoveUnsupportedLayouts(
+    smart_objects::SmartObject& keyboard_capabilities,
+    const std::set<mobile_apis::KeyboardLayout::eType>& supported_layouts,
+    std::set<mobile_apis::KeyboardLayout::eType>& customizable_layouts) {
+  SDL_LOG_AUTO_TRACE();
+
+  if (!keyboard_capabilities.keyExists(hmi_response::configurable_keys)) {
+    SDL_LOG_WARN(
+        "No configurable keys param in received keyboard capabilities");
+    return;
+  }
+
+  auto configurable_keyboards =
+      keyboard_capabilities[hmi_response::configurable_keys].asArray();
+  if (!configurable_keyboards) {
+    SDL_LOG_WARN("Pointer to configurable_keyboards array is null");
+    return;
+  }
+
+  for (auto keyboard = (*configurable_keyboards).begin();
+       keyboard != (*configurable_keyboards).end();) {
+    auto found_layout =
+        supported_layouts.find(static_cast<mobile_apis::KeyboardLayout::eType>(
+            (*keyboard)[hmi_request::keyboard_layout].asInt()));
+    if (supported_layouts.end() == found_layout) {
+      (*configurable_keyboards).erase(keyboard);
+    } else {
+      customizable_layouts.insert(
+          static_cast<mobile_apis::KeyboardLayout::eType>(
+              (*keyboard)[hmi_request::keyboard_layout].asInt()));
+      ++keyboard;
+    }
+  }
+}
+
+/**
+ * @brief AddMissedSupportedLayouts adds missed supported layouts to
+ * configurable_keys array, if present. If configurable_keys param is absent, it
+ * will be created with all supported layouts and default number of keys (zero).
+ * @param keyboard_capabilities Modified keyboard capabilities
+ * @param supported_layouts Set with all supported layouts, generated from
+ * supported_keyboard_layouts array
+ * @param customizable_layouts Set with all supported layouts, which keys may be
+ * customized (if not, number of keys for such layout will be set to zero by
+ * default)
+ */
+void AddMissedSupportedLayouts(
+    smart_objects::SmartObject& keyboard_capabilities,
+    const std::set<mobile_apis::KeyboardLayout::eType>& supported_layouts,
+    const std::set<mobile_apis::KeyboardLayout::eType>& customizable_layouts) {
+  for (auto layout : supported_layouts) {
+    auto found_layout = customizable_layouts.find(layout);
+
+    if (customizable_layouts.end() == found_layout) {
+      smart_objects::SmartObject missed_capabilities(
+          smart_objects::SmartType_Map);
+      missed_capabilities[hmi_response::num_configurable_keys] = 0;
```

Comment:
@ychernysheva I would say, `default_capabilities`

Reply:
Done in a9c3f0b